flink-runtime/src/main/java/org/apache/flink/runtime/operators/hash/ReduceHashTable.java

```diff
@@ -0,0 +1,1014 @@
+	private final StagingOutputView stagingSegmentsOutView;
+
+	private T reuse;
+
+	/** This is the internal prober that insertOrReplaceRecord and processRecordWithReduce use. */
+	private final HashTableProber<T> prober;
+
+	/** The number of elements currently held by the table. */
+	private long numElements = 0;
+
+	/** The number of bytes wasted by updates that couldn't overwrite the old record. */
+	private long holes = 0;
+
+	/**
+	 * If the serializer knows the size of the records, then we can calculate the optimal number of buckets
+	 * upfront, so we don't need resizes.
+	 */
+	private boolean enableResize;
+
+
+	/**
+	 * This constructor is for the case when will only call those operations that are also
+	 * present on CompactingHashTable.
+	 */
+	public ReduceHashTable(TypeSerializer<T> serializer, TypeComparator<T> comparator, List<MemorySegment> memory) {
+		this(serializer, comparator, memory, null, null, false);
+	}
+
+	public ReduceHashTable(TypeSerializer<T> serializer, TypeComparator<T> comparator, List<MemorySegment> memory,
+						ReduceFunction<T> reducer, Collector<T> outputCollector, boolean objectReuseEnabled) {
+		super(serializer, comparator);
+		this.reducer = reducer;
+		this.numAllMemorySegments = memory.size();
+		this.freeMemorySegments = new ArrayList<>(memory);
+		this.outputCollector = outputCollector;
+		this.objectReuseEnabled = objectReuseEnabled;
+
+		// some sanity checks first
+		if (freeMemorySegments.size() < MIN_NUM_MEMORY_SEGMENTS) {
+			throw new IllegalArgumentException("Too few memory segments provided. ReduceHashTable needs at least " +
+				MIN_NUM_MEMORY_SEGMENTS + " memory segments.");
+		}
+
+		// Get the size of the first memory segment and record it. All further buffers must have the same size.
+		// the size must also be a power of 2
+		segmentSize = freeMemorySegments.get(0).size();
+		if ( (segmentSize & segmentSize - 1) != 0) {
+			throw new IllegalArgumentException("Hash Table requires buffers whose size is a power of 2.");
+		}
+
+		this.numBucketsPerSegment = segmentSize / bucketSize;
+		this.numBucketsPerSegmentBits = MathUtils.log2strict(this.numBucketsPerSegment);
+		this.numBucketsPerSegmentMask = (1 << this.numBucketsPerSegmentBits) - 1;
+
+		recordArea = new RecordArea(segmentSize);
+
+		stagingSegments = new ArrayList<>();
+		stagingSegmentsInView = new RandomAccessInputView(stagingSegments, segmentSize);
+		stagingSegmentsOutView = new StagingOutputView(stagingSegments, segmentSize);
+
+		prober = new HashTableProber<>(buildSideComparator, new SameTypePairComparator<>(buildSideComparator));
+
+		enableResize = buildSideSerializer.getLength() == -1;
+	}
+
+	private void open(int numBucketSegments) {
+		synchronized (stateLock) {
+			if (!closed) {
+				throw new IllegalStateException("currently not closed.");
+			}
+			closed = false;
+		}
+
+		allocateBucketSegments(numBucketSegments);
+
+		stagingSegments.add(allocateSegment());
```

Comment:
This shouldn't happen, because the size of the bucket area is determined to never encompass all the segments we have. (Note that at the beginning of `open`, all segments should be free.)

But nevertheless I'll add a check and throw a "This shouldn't happen" RuntimeException.